# PAPST: Peak Assignment and Profile Search Tool

A tool for Chip-seq analysis.

![screen shot](https://github.com/paulbible/papst/raw/master/tutorials/papst_screen.png)

## Post Peak-calling Analysis

PAPST is a user friendly program for analysis of ChIP-Seq (chromatin immunoprecipitation followed by sequencing) experiments.   

## Features
The main features of PAPST are:
* Transcription Factor (TF) and Epigenetic Mark (EM) profile pattern search
* Assignment of ChIP-seq peaks of known genes.
* Analysis of peak binding distribution (% promoter, % gene body, etc.)
* Calculate shared/overlapping peaks between multiple ChIP-seq experiments
* Genome agnostic analysis, supports all genomes available through UCSC Genome Browser
* Quickly and interactively analyze ChIP-Seq peaks from multiple experiments

## Getting Started
The [tutorials folder](https://github.com/paulbible/papst/tree/master/tutorials) above contains detailed walkthroughs with images for both Mac and Windows. Test data is available in the data folder above.

## Requirements
The only requirement is Java 1.7. The lastest version of Java is available at http://java.com/en/download/.

## License
PAPST is released into the public domain and free for any use. See DISCLAIMER.

## Releases
The current version of PAPST is available under [releases](https://github.com/paulbible/papst/releases).

## How to Cite
Check back later.

## Acknowledgments
This work was supported by the Intramural Research Program (IRP) and the Office of Science and Technology (OST) of the National Institute of Arthritis and Musculoskeletal and Skin Diseases (NIAMS) of the National Institutes of Health (NIH).

PAPST used public domain images from https://openclipart.org/.



